Support pinning of transitive dependencies for central package version managed enabled projects.

cristinamanum opened the pull request:
## Feature

https://github.com/NuGet/Home/issues/8910

Regression: No

## Fix

Transitive dependencies for Package Versions defined centrally will need to be respected in the dependency graph. 
This PR adds support pinning of transitive dependencies for CPVM enabled projects.

To enable this feature the central package versions that are transitive dependencies will be added to the restore graph as top level nodes like they were direct dependencies. With this approach the logic for the graph analysis stays at it is. 
Step by step logic
1. The PackageSpecReferenceDependencyProvider is updated to be aware of the central package versions and include them on the  GetSpecDependencies call but marked as AutoReferenced and CentrallyManaged. The two flags are enough to mark them that they are not added by the user and they are centrally managed. These settings will be used to filter them distinct from the other dependencies in the list.
2. The RemoteDependencyWalker during the graph creation will not add them(the above central auto referenced dependencies) by default but only if they will be transitive. 
At the first pass if a transitive dependency has the version that is a central one it will marked and later added to the graph. To achieve this the internal TransitiveCentralPackageVersions type was introduced to facilitate the marking and consumption of the transitive central dependencies. The internal ConcurrentDictionary was used in order to avoid an implementation for a concurrent HashSet. 
3. The downgrades will be marked as errors in case of projects using CPVM. 

## Testing/Validation

Tests Added: Yes

Validation:  Unit tests and manual E2E tests.

cristinamanum committed:
Add the support for central package versions and transitive dependencies.

cristinamanum committed:
Fix unit test

cristinamanum committed:
Style updates

cristinamanum committed:
Bug fix

cristinamanum committed:
Bug fix.

cristinamanum committed:
Remove extra line

cristinamanum committed:
Merge branch 'dev' into dev-cmanu-transitivedep

cristinamanum committed:
Test fix

cristinamanum commented on test/TestUtilities/Test.Utility/DependencyResolver/DependencyProvider.cs:
```diff
@@ -0,0 +1,169 @@
+
+namespace Test.Utility
+{
+    public class DependencyProvider : IRemoteDependencyProvider, IDependencyProvider
```
This code is moved from NuGet.DependencyResolver.Core.Tests/RemoteDependencyWalkerTests.cs .

dominoFire commented on src/NuGet.Core/NuGet.Commands/Strings.resx:
```diff
@@ -973,4 +973,7 @@ For more information, visit https://docs.nuget.org/docs/reference/command-line-r
   </data>
-</root>
+  <data name="Log_CPVM_DowngradeError" xml:space="preserve">
+    <value>Detected package downgrade: {0} from {1} to centrally defined {2}.</value>
```
Add a description to the placeholders ({0}, {1} ...) to aid the translation team.

cristinamanum replied:
Updated.

nkolev92 commented:
@cristinamanum 

Sorry for the delay in reviewing this. 

Can you rebase and resolve the conflict?
Additionally, can you please update the title/add a quick description for the PR as the changes are in a relatively stable part of restore. It'll make it quicker to review things.

zivkan commented:
+1 to Nikolche's request for some description. I've never looked at the resolver code before, so I have no idea how it actually works. THis is the 1st PR to modify it in over 12 months.

I know NuGet selects the lowest version that is greater than or equal to the min version in the version range, so I imagined a method something like `NuGetVersion SelectVersionForPackage(string packageId, VersionRange versionRange)`. I hoped implementation could just look up the package in an immutable collection of centrally defined package versions, and only ask `NuGet.Protocol` for a list of versions available on each source when no version was found defined centrally. something like:

```cs
NuGetVersion SelectVersionForPackage(string packageId, VersionRange versionRange)
{
  if (CentrallyManagedPackageVersions.TryGet(packageId, out var version))
  {
    return version;
  }
  else
  {
    // whatever the current code does.
    return GetAllVersionsFromAllSources(packageId)
      .Where(p => p.Version > versionRange.MinVersion)
      .Min();
  }
}
```

So, I don't really understand why there's a mutable queue of central package versions to be processed. I found the dictionary of bools confusing as well, but it appears you may not have been aware of `HashSet<T>`, and are using a dictionary as something functionally equivalent.

dominoFire commented on src/NuGet.Core/NuGet.Common/Errors/NuGetLogCode.cs:
```diff
@@ -244,6 +244,11 @@ public enum NuGetLogCode
+        /// <summary>
+        /// Package dependency downgraded because of centrally defined package version.
+        /// </summary>
+        NU1609 = 1609,
```
If you are using this log code for errors, you should move to a number between 1100-1499, as stated at the beginning of the file:

https://github.com/NuGet/NuGet.Client/blob/ee128e447232eb66c71598cb8a335d6868eec266/src/NuGet.Core/NuGet.Common/Errors/NuGetLogCode.cs#L27-L33

cristinamanum replied:
Changed to NU1109.

dominoFire commented on test/NuGet.Core.Tests/NuGet.ProjectModel.Test/PackageSpecReferenceDependencyProviderTests.cs:
```diff
@@ -0,0 +1,97 @@
+                var fooDep = dependencies.Where(d => d.Name == "foo").First();
+                Assert.NotNull(fooDep);
+                Assert.False(fooDep.VersionCentrallyManaged);
+                Assert.False(fooDep.VersionCentrallyManaged);
```
Do you need to assert twice?

cristinamanum replied:
No - removed one Assert.

dominoFire commented on test/NuGet.Core.Tests/NuGet.ProjectModel.Test/PackageSpecReferenceDependencyProviderTests.cs:
```diff
@@ -0,0 +1,97 @@
+        {
+            // Arrange
+            var logger = new TestLogger();
+            var dependencyFoo = new LibraryDependency(new LibraryRange("foo", versionRange: null, LibraryDependencyTarget.Package),
```
Complete the named arguments

cristinamanum replied:
Updated.

cristinamanum committed:
Dev merge

zivkan commented:
> Two enable this feature the central package versions that are transitive dependencies will be added to the restore graph as top level nodes like they were direct dependencies. With this approach the logic for the graph analysis stays at it is.

That's a neat trick. Will it change the graph that's written to the assets file?  If so, then I imagine this will make `pack` automatically get the pinned versions as well. It's been a long time since I looked at the CPVM spec. Is that desired?

dominoFire commented on test/TestUtilities/Test.Utility/DependencyResolver/DependencyProvider.cs:
```diff
@@ -0,0 +1,169 @@
+            {
+                return false;
+            }
+        }
```
```suggestion
        public bool IsHttp => false;
```

cristinamanum replied:
Updated.

dominoFire commented on test/NuGet.Core.Tests/NuGet.DependencyResolver.Core.Tests/RemoteDependencyWalkerTests.cs:
```diff
@@ -823,176 +825,202 @@ public void IsGreaterThanEqualTo_ReturnsFalse_IfRightVersionIsLargerThanLeft(str
 
-        private Task<GraphNode<RemoteResolveResult>> DoWalkAsync(RemoteDependencyWalker walker, string name)
+        [Fact]
+        public void TransitiveCentralPackageVersions__TryAdd_DuplicatesAreIgnored()
```
```suggestion
        public void TransitiveCentralPackageVersions_TryAdd_DuplicatesAreIgnored()
```

cristinamanum replied:
Updated.

cristinamanum commented:
> @cristinamanum
> 
> Sorry for the delay in reviewing this.
> 
> Can you rebase and resolve the conflict?
> Additionally, can you please update the title/add a quick description for the PR as the changes are in a relatively stable part of restore. It'll make it quicker to review things.

Updated the summary. title and fixed the conflicts.

cristinamanum commented:
> > Two enable this feature the central package versions that are transitive dependencies will be added to the restore graph as top level nodes like they were direct dependencies. With this approach the logic for the graph analysis stays at it is.
> 
> That's a neat trick. Will it change the graph that's written to the assets file? If so, then I imagine this will make `pack` automatically get the pinned versions as well. It's been a long time since I looked at the CPVM spec. Is that desired?

The pack new behavior is a required part of the spec. There are still changes needed for that. A new PR will come after this.

There will not be changes to the assets file with this change. It will be with the Pack needed changes.  The assets file will reflect the packages versions resolved/downloaded. If PackA -> PackB (1.0.0)  and PackB (2.0.0) is central the  PackB 2.0.0 will be in the assets.

dominoFire reviewed (commented):
Overall looks good. Please address the log code comment.

cristinamanum committed:
Feedback

nkolev92 commented on src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreCommand.cs:
```diff
@@ -270,7 +271,7 @@ public async Task<RestoreResult> ExecuteAsync(CancellationToken token)
                 using (TelemetryActivity.CreateTelemetryActivityWithNewOperationIdAndEvent(parentId: _operationId, eventName: ValidateRestoreGraphs))
                 {
-                    _success &= await ValidateRestoreGraphsAsync(graphs, _logger);
+                    _success &= await ValidateRestoreGraphsAsync(graphs, _logger, isCpvmEnabled);
```
nit: You might not need to pass the extra parameter around, it's part of the restoreRequest which is a property in this class. 
At least not in the first method. It's already commonly used in other places.

Something to consider, marking it as a nit, this class needs some work, it's gotten messy lately :)

cristinamanum replied:
Removed the parameter,

nkolev92 commented on src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreCommand.cs:
```diff
@@ -815,13 +816,20 @@ private static Task LogDowngradeWarningsAsync(IEnumerable<RestoreTargetGraph> gr
-                            messages.Add(RestoreLogMessage.CreateWarning(NuGetLogCode.NU1605, message, downgraded.Key.Name, graph.TargetGraphName));
+                            if (isCpvmEnabled)
+                            {
+                                messages.Add(RestoreLogMessage.CreateError(NuGetLogCode.NU1109, message, downgraded.Key.Name, graph.TargetGraphName));
```
How do you differentiate between a CPVM caused transitive downgrade and a regular downgrade?

cristinamanum replied:
Currently all the downgrades for CPVM are errors. My thought was to not differentiate between the types of dependencies. Why would be a differentiation?

zivkan replied:
Because "normal" downgrade errors are resolved by adding a direct reference to the package. CPVM downgrades are resolved possibly by adding a PackageVersion to the package, if one doesn't already exist, or upgrading the version if a PackageVersion does already exist for the package. They have potentially different causes and resolutions.

Is transient version pinning always-on, or is it another opt-in feature? If it's opt-in, that's another reason to detect it differently.

nkolev92 replied:
Just adding here that I agree with Andy.
The resolution method is different, so the messages should be slightly different as well. 
In 1605 we say reference the package directly,
> Reference the package directly from the project to select a different version.

In CPVM we should say that the package version should be lifted.

cristinamanum replied:
Updated the code to differentiate.

cristinamanum replied:
Updated the code to differentiate.

nkolev92 commented on src/NuGet.Core/NuGet.ProjectModel/PackageSpecReferenceDependencyProvider.cs:
```diff
@@ -326,6 +326,19 @@ private List<LibraryDependency> GetDependenciesFromSpecRestoreMetadata(PackageSp
                 var targetFrameworkInfo = packageSpec.GetTargetFramework(targetFramework);
                 dependencies.AddRange(targetFrameworkInfo.Dependencies);
 
+                if (packageSpec.RestoreMetadata?.CentralPackageVersionsEnabled ?? false)
```
```suggestion
                if (packageSpec.RestoreMetadata?.CentralPackageVersionsEnabled == true)
```

Positive conditions read a bit better.

cristinamanum replied:
Updated.

nkolev92 commented on src/NuGet.Core/NuGet.ProjectModel/PackageSpecReferenceDependencyProvider.cs:
```diff
@@ -326,6 +326,19 @@ private List<LibraryDependency> GetDependenciesFromSpecRestoreMetadata(PackageSp
+                        {
+                            LibraryRange = new LibraryRange(item.Value.Name, item.Value.VersionRange, LibraryDependencyTarget.Package),
+                            VersionCentrallyManaged = true,
+                            AutoReferenced = true,
```
Why in particular do we need this?

If we need a different concept, I think we should add it. 
Autoreferenced is being misused right now.

cristinamanum replied:
Added a different concept.

cristinamanum replied:
I missed to answer to this: "Why in particular do we need this?" 

The provider will be aware of the central package versions but these dependencies needs to not be like the direct ones, they will be "weaker". Their job is to be there in case that a transitive version needs to be updated by a central version and also to catch any downgrades. When during the graph creations the provider is asked to provide the dependencies (https://github.com/NuGet/NuGet.Client/blob/71e16df95a76ce4ff624ecb7c912a479c33c1da9/src/NuGet.Core/NuGet.DependencyResolver.Core/Remote/RemoteDependencyWalker.cs#L145) the "weaker" dependencies will not be added to the graph. They will be added only as needed (== if a transitive dependency is present in the central list).

nkolev92 commented on test/NuGet.Core.Tests/NuGet.Commands.Test/RestoreCommandTests.cs:
```diff
@@ -1416,6 +1417,55 @@ public async Task RestoreCommand_CentralVersion_ErrorWhenCentralPackageVersionFi
         }
 
+        [Fact]
+        public async Task RestoreCommand_DowngradeIsErrorWhenCpvmEnabled()
```
This is not really a RestoreCommand test. 
It's a merely testing the resolver. 

I think you should add a proper RestoreCommandTest. 

I have some samples for RestoreCommand set-ups in https://github.com/NuGet/NuGet.Client/compare/dev-nkolev92-addRestoreResolverTests#diff-d07676aab31eda417a4b32f5935402c0R1349 for reference. 

I will try to get some of those tests in later.

nkolev92 replied:
It'd also be helpful if the graph is "drawn out" somewhere in the test, rather than explained.

cristinamanum replied:
I tried to make a picture to explain better. I added a new test based on the sample provided. I still let the old test.

nkolev92 commented on src/NuGet.Core/NuGet.DependencyResolver.Core/Remote/RemoteDependencyWalker.cs:
```diff
@@ -155,17 +174,25 @@ public Task<GraphNode<RemoteResolveResult>> WalkAsync(LibraryRange library, NuGe
                     }
                     else
                     {
+                        // In case of conflict because of a centrally managed version
```
Can you elaborate here? 
Is there a test capturing this scenario that I can debug into?

cristinamanum replied:
Few tests:
RestoreCommanTests
RestoreCommand_DowngradeIsNotErrorWhen_DowngradedByCentralDirectDependency
RestoreCommand_DowngradeIsErrorWhen_DowngradedByCentralTransitiveDependency

RemoteDependencyWalkerTests
WalkAsyncAddsTransitiveCentralDependency

cristinamanum committed:
Add more tests, force downgrade error only when forced by central transitive; other feedback.

cristinamanum committed:
Fix tests

cristinamanum committed:
Test fix

zivkan commented on src/NuGet.Core/NuGet.DependencyResolver.Core/Remote/RemoteDependencyWalker.cs:
```diff
@@ -136,10 +155,10 @@ public Task<GraphNode<RemoteResolveResult>> WalkAsync(LibraryRange library, NuGe
                     if (StringComparer.OrdinalIgnoreCase.Equals(dependency.Name, libraryRange.Name))
                     {
-                        result = DependencyResult.Cycle;
+                        result = (DependencyResult.Cycle, dependency);
```
I think it would be a bit easier to understand if you use [ValueTuple deconstruction](https://docs.microsoft.com/en-us/dotnet/csharp/deconstruct#deconstructing-a-tuple). Maybe `result` isn't a good name and needed to be changed anyway, but if it doesn't, they it would have avoided changing all the usage of it to include `.dependencyResult`

cristinamanum replied:
Updated with named tuples.

zivkan replied:
That's not what I was trying to suggest.

you changed the predicate so `var result = predicate(dependency.Library)` no longer results a single enum value, it returns a ValueTuple. As a result, every time `result` was read, for example `if (result == something || result == somethingElse)`, you had to change it to `if (result.dependencyResult == something || result.dependencyResult == somethingElse)`.

If instead you used value tuple deconstruction, `var (result, conflictingDependency) = predicate(dependencyLibrary)`, then you could have left that `if (result == something || result == somethingElse)` unmodified. And when you want to read `conflictingDependency` you wouldn't need `result.conflictingDependency`.

zivkan commented on src/NuGet.Core/NuGet.DependencyResolver.Core/Remote/RemoteDependencyWalker.cs:
```diff
@@ -343,5 +372,82 @@ private enum DependencyResult
+        private void MarkCentralVersionForTransitiveProcessing(NuGetFramework framework, LibraryDependency libraryDependency, TransitiveCentralPackageVersions transitiveCentralPackageVersions)
+        {
+            // Set the transitive flag once is marked as transitive
+            libraryDependency.ReferenceType = LibraryDependencyReferenceType.Transitve;
```
The comment doesn't help understanding. It looks like it's just describing what happens on the following line.

cristinamanum replied:
Removed the comment as it does not bring value.

zivkan commented on src/NuGet.Core/NuGet.DependencyResolver.Core/Remote/RemoteDependencyWalker.cs:
```diff
@@ -343,5 +372,82 @@ private enum DependencyResult
+        {
+            return dependency.ReferenceType == LibraryDependencyReferenceType.Direct ||
+                   !dependency.VersionCentrallyManaged;
+        }
```
I think it would be easier to understand if centrally managed versions were added with `LibraryDepdendencyReferenceType.None`, so that this check would be nothing more than `dependency.ReferenceType != LDR.None`.

Or add a new referenced type `Unused`.


Would that also remove the need for the `TransitiveCentralPackageVersions` class? You could just iterate the dependencies and find the ones where `VersionCentrallyManaged == true && ReferenceType == LDR.Transitive`

cristinamanum replied:
The dependencies that need to be ignored are set with LDR.None, but I see your point.  I changed the code to be more explicit. This will not remove the need for the TransitiveCentralPackageVersions. The dependencies that needs to be processed are found at the graph traversal time.

zivkan reviewed (commented):
I'm going for a break, but here are my comments so far.

dominoFire commented on src/NuGet.Core/NuGet.ProjectModel/PackageSpecReferenceDependencyProvider.cs:
```diff
@@ -326,6 +326,19 @@ private List<LibraryDependency> GetDependenciesFromSpecRestoreMetadata(PackageSp
                 var targetFrameworkInfo = packageSpec.GetTargetFramework(targetFramework);
                 dependencies.AddRange(targetFrameworkInfo.Dependencies);
 
+                if (packageSpec.RestoreMetadata?.CentralPackageVersionsEnabled == true)
```
```suggestion
                if (packageSpec.RestoreMetadata?.CentralPackageVersionsEnabled ?? false)
```

dominoFire replied:
I just read the previous comment in this line. No problem at all.

nkolev92 replied:
This was feedback from me earlier. 
Always use positive conditions wherever possible. They are easier to read.

cristinamanum replied:
Changed.

dominoFire commented on test/NuGet.Core.Tests/NuGet.Commands.Test/RestoreCommandTests.cs:
```diff
@@ -1416,6 +1417,216 @@ public async Task RestoreCommand_CentralVersion_ErrorWhenCentralPackageVersionFi
+
+                packageA_Version200.Dependencies.Add(packageB_Version200);
+
+
```
https://github.com/NuGet/NuGet.Client/blob/dev/docs/coding-guidelines.md Rule 7:

> Avoid more than one empty line at any time. For example, do not have two blank lines between members of a type.

cristinamanum replied:
Removed the extra line.

dominoFire commented on test/NuGet.Core.Tests/NuGet.Commands.Test/RestoreCommandTests.cs:
```diff
@@ -1434,5 +1645,17 @@ private static TargetFrameworkInformation CreateTargetFrameworkInformation(List<
+            };
+
+            return walker.WalkAsync(range, framework, runtimeIdentifier: null, runtimeGraph: null, recursive: true);
+
```
Looks like a extra empty line

cristinamanum replied:
Yes - removed.

dominoFire commented on test/NuGet.Core.Tests/NuGet.DependencyResolver.Core.Tests/RemoteDependencyWalkerTests.cs:
```diff
@@ -823,176 +818,246 @@ public void IsGreaterThanEqualTo_ReturnsFalse_IfRightVersionIsLargerThanLeft(str
 
-        private void AssertPath<TItem>(GraphNode<TItem> node, params string[] items)
+        [Fact]
+        public void TransitiveCentralPackageVersions__AddAndTake()
```
```suggestion
        public void TransitiveCentralPackageVersions_AddAndTake()
```

cristinamanum replied:
Updated.

dominoFire commented on test/NuGet.Core.Tests/NuGet.DependencyResolver.Core.Tests/RemoteDependencyWalkerTests.cs:
```diff
@@ -823,176 +818,246 @@ public void IsGreaterThanEqualTo_ReturnsFalse_IfRightVersionIsLargerThanLeft(str
+        private Task<GraphNode<RemoteResolveResult>> DoWalkAsync(RemoteDependencyWalker walker, string name)
+        {
+            return DoWalkAsync(walker, name, NuGetFramework.Parse("net45"));
 
```
Looks like an extra line.

cristinamanum replied:
Yes it was. Removed.

dominoFire commented on test/NuGet.Core.Tests/NuGet.DependencyResolver.Core.Tests/RemoteDependencyWalkerTests.cs:
```diff
@@ -823,176 +818,246 @@ public void IsGreaterThanEqualTo_ReturnsFalse_IfRightVersionIsLargerThanLeft(str
+        public void IsDependencyValidForGraphTest(LibraryDependencyReferenceType referenceType, bool versionCentrallyManaged)
+        {
+            var centralPackageName = "D";
+            var framework = NuGetFramework.Parse("net45");
```
Looks like this variable is not used in this test.

cristinamanum replied:
Removed.

dominoFire commented on src/NuGet.Core/NuGet.ProjectModel/PackageSpecReferenceDependencyProvider.cs:
```diff
@@ -326,6 +326,19 @@ private List<LibraryDependency> GetDependenciesFromSpecRestoreMetadata(PackageSp
+                        {
+                            LibraryRange = new LibraryRange(item.Value.Name, item.Value.VersionRange, LibraryDependencyTarget.Package),
+                            VersionCentrallyManaged = true,
+                            ReferenceType = LibraryDependencyReferenceType.None,
```
In line 332, you're adding central dependencies that are not already in the `depenencies` list. Are those central dependencies considered transitive ? If that's the case, should the ReferenceType be `LibraryDependencyReferenceType.Transitve` ?

cristinamanum replied:
No, at this point they are not transitive for this reason set to none.

dominoFire reviewed (approved):
Looks good. Style comments and a question on PackageSpecReferenceDependencyProvider. That should not block this PR.

cristinamanum committed:
Feedback

zivkan commented on src/NuGet.Core/NuGet.DependencyResolver.Core/Remote/RemoteDependencyWalker.cs:
```diff
@@ -24,18 +25,34 @@ public RemoteDependencyWalker(RemoteWalkContext context)
+                outerEdge: null,
+                transitiveCentralPackageVersions: transitiveCentralPackageVersions);
+
+            while (transitiveCentralPackageVersions.TryTake(out LibraryDependency centralPackageversionDependecy))
+            {
+                await AddTransitiveCentralPackageVersionNodesAsync(rootNode, centralPackageversionDependecy, framework, runtimeIdentifier, runtimeGraph, transitiveCentralPackageVersions);
+            }
```
Here's an alternative that doesn't need the `TransitiveCentralPackageVersions` inner class.

What do you think?

If performance is a concern, the innerNodes foreach loop can check that the node is a project, not a package, before pushing it onto the stack.

```suggestion
            var nodesToCheck = new Stack<GraphNode<RemoteResolveResult>>();
            nodesToCheck.Push(rootNode);

            while (nodesToCheck.Count > 0)
            {
                var node = nodesToCheck.Pop();

                foreach (var dependency in node.Item.Data.Dependencies)
                {
                    if (dependency.VersionCentrallyManaged && dependency.ReferenceType == LibraryDependencyReferenceType.Transitve)
                    {
                        await AddTransitiveCentralPackageVersionNodesAsync(rootNode, dependency, framework, runtimeIdentifier, runtimeGraph);
                    }
                }

                foreach (var innerNode in node.InnerNodes)
                {
                    nodesToCheck.Push(innerNode);
                }
            }
```

cristinamanum replied:
> Here's an alternative that doesn't need the `TransitiveCentralPackageVersions` inner class.
> 
> What do you think?
> 
> If performance is a concern, the innerNodes foreach loop can check that the node is a project, not a package, before pushing it onto the stack.

Couple of observations regarding the suggestion:
1. There is not any need to loop through the entire dependency list in node.Item.Data.Dependencies.
2. There should not be duplicates in the nodes. With the above suggestion are going to be. 

Observation about the current implementation:
The current `TransitiveCentralPackageVersions` can be thought as concurrent queue that does not allow duplicate enqueues. The items are added during the graph traversal with no need to traverse again the inner nodes. 
What do you think is incorrect about the current approach?

zivkan replied:
> 1. There is not any need to loop through the entire dependency list in node.Item.Data.Dependencies.

There is if you don't keep a collection of used centrally managed transitive dependencies, which was the entire point of my suggestion. At least, that's the only place I could find ReferenceType to check if it's value is `None` or `Transitive` (find which ones were used, ignore the ones not used).

> 2. There should not be duplicates in the nodes. With the above suggestion are going to be.

In what scenario will there be duplicates? I thought the `PackageVersion` items are de-duplicated before restore, just as PackageReference and PackageDownload items are, so by the time it gets to this method there are no duplicate items. Is it because if a CPVM project has a project reference to another CPVM project, that child project is going to have its own items, which could be duplicates?

This makes me curious if there's a bug in both of our implementations: Should centrally managed versions of project references even be in the current project's restore graph? In particular, if the other project uses a different `Directory.Packages.props` file with different packages and versions defined, would we even want those other proejct's `PackageVersion` items from appearing, and therefore changing, the current project's restore graph? I haven't yet had a look at your tests, but you can answer this question quickly: Do you have a test where two projects use different `Directory.Packages.props` files that define different versions for packages from each other, and one of the projects has a `ProjectReference` to the other?  What's the expected behaviour, particularly in edge cases when the project being restored does not have a `PackageVersion` defined in its `DPp` file, but the referenced project does, which is different to the version defined in the package's dependencies list? What about when only one of the two projects are CPVM opt-in?

> What do you think is incorrect about the current approach?

Nothing incorrect (Except maybe the case I just described above, where project references use different CPVM files), but I found it difficult to understand. Before debugging to make my proposal I didn't know anything at all about this code, and that's where most of my difficulty understanding comes from. But when I see someone use `Concurrent` classes, to do a good job reviewing, I have to ask myself, why is a concurrent data structure used? So, I spent time to find out that dependencies are processed in parallel. Ok, since it's in parallel and modifying a single object's instance, it's good that concurrent data structures are used, but are they used correctly? Is there a potential parallel timing bug? Some people are big fans of lock-free concurrent algorithms, because it's easier to understand and less prone to bugs, because it's hard to do parallel/concurrent programming right. Even with all the APIs and data structures, highly skilled developers still make mistakes. So, even if you've done it right, I feel obligated to do a more in-depth review to make sure I think it's correct.

Additionally, not only did you use the `ConcurrentQueue` data structure in the implementation, you also described the change as "can be thought as concurrent queue". In my mind, queues have the characteristic of being a first-in, first-out collection, and is typically used in consumer-producer patterns. So now I'm asking myself why does the order matter? Why are things being removed in the queue while other tasks/threads are still adding things to the queue? As far as I can tell, that's not the case. It's just getting the set of items that were "promoted" from not used to a transitive dependency and the semantics are "write-only until graph walking is complete, then read-only". You could have achieved the same thing using only a `ConcurrentDictionary`, since there is no `ConcurrentSet`, and then iterating the dictionary's keys.

Maybe I'm overthinking it, but this is how I do code reviews, and why it takes me a long time. Something else I try to consider is if someone else is going to modify this code in 12 months from now. Will the code be easy to understand and modify?

My high level description of your implementation is:
* Recursively walk the dependencies graph
  * When a project is evaluated, add fake dependencies for all the package versions that do not already have a direct dependency.
  * When a "conflict" is found, and it's for one of these fake dependencies, mark the dependency as used and add it to a collection of used transitive dependencies.
* When graph walking is finished, get the used dependencies from the collection, and add them as real dependencies in the graph.

My proposal's high level description is like this:
* Recursively walk the dependencies graph
  * When a project is evaluated, add fake dependencies for all the package versions that do not already have a direct dependency.
  * When a "conflict" is found, and it's for one of these fake dependencies, mark the dependency as used.
* When graph walking is finished, find the used fake dependencies and add them as real dependencies.

When writing it out like this, it gave me a different idea:
* Don't add fake dependencies, and do a graph walk to download the full dependency graph, just like non-CPVM.
* Enumerate the graph to find all dependencies, and when the package ID matches a pinned package, add a top level dependency for it.

If the scenario at the top of this ridiculously long comment is really a bug (project reference with different CPVM files), not only will this last idea be easier for people to understand, and possibly have simpler code as a result, it won't be susceptible to the bug.

cristinamanum replied:
I will try to address the points in the comment above. 
1. I do have a set of E2E tests used to validate scenarios like the ones mentioned. In case the Project1 ->Project2. Both use CPVM and different versions in the two Directory.Packages.props the main project dependencies wins. 
2. The two logic sequences above are similar. In the current implementation the graph is not traversed multiple times. The dependencies that need to be added are marked at the moment the graph nodes are created using this new type. In the other proposal the new type is not needed, it is eventually replaced with a Concurrent Collection but reiterate multiple times through the graph.
3. It was a comment about duplication. Imagine 
Project1 -> PackageA -> PackageD -> PackageB
                                    -> PackageC -> PackageB
Central PackageB
In this case the PackageB should be only marked once to be added to the graph. 
4. I am not sure if I understated well the comments regarding the concurrent need for the `TransitiveCentralPackageVersions` - the adds are done from different threads. 
5. There was a comment regarding the ordering. It could be very well a ConcurrentBag ( that is implemented on top of ConcurrentQueue). I mentioned it only trying to explain more the logic. The `TransitiveCentralPackageVersions` has only two methods (TryAdd, TryTake) none of them implying an order. 
6. I understand that the new changes are difficult to understand. I will add more comments to make it more clear.

nkolev92 replied:
I just wanted to add a specific emphasis to one of @zivkan's comments that I really agree with: 

> Don't add fake dependencies, and do a graph walk to download the full dependency graph, just like non-CPVM.

fwiw, I think the fake dependency is confusing. 
It's a separate concept that's hard to follow right now. 
The fact that the library range can now represent a transitive dependency is confusing.  
The central dependencies could be a only one object pasted to the walker when appropriate.

cristinamanum committed:
Fix unit tests

nkolev92 commented on src/NuGet.Core/NuGet.Commands/Strings.resx:
```diff
@@ -980,4 +980,8 @@ For more information, visit https://docs.nuget.org/docs/reference/command-line-r
     <comment>don't localize 'format', 'detailed' or 'short'</comment>
   </data>
+  <data name="Log_CPVM_DowngradeError" xml:space="preserve">
+    <value>Detected package downgrade: {0} from {1} to centrally defined {2}.</value>
```
We are not calling out the action here like in the way that `Strings.Log_DowngradeWarning` does. 

The other one says: 

> Reference the package directly from the project to select a different version. 

This one should probably say 

> Update the centrally managed package version to a higher version.

We can even call out the exact version, but that could be tricky because 2 different projects might have 2 different graphs and the versions needed to fix them both might be different.

cristinamanum replied:
Updated

nkolev92 commented on src/NuGet.Core/NuGet.DependencyResolver.Core/GraphModel/GraphItem.cs:
```diff
@@ -18,19 +19,49 @@ public GraphItem(LibraryIdentity key)
+        /// A <see cref="LibraryDependency"/> created from a <see cref="CentralPackageVersion"/>.
+        /// If set it will indicate that the graph node was created due to a transitive dependency for a package that was also defined in the central packages management file. 
+        /// </summary>
+        public LibraryDependency CentralDependency { get; set; }
```
Can this be a bool then?

cristinamanum replied:
Not really. This will be used in pack functionality.

nkolev92 commented on src/NuGet.Core/NuGet.LibraryModel/LibraryDependency.cs:
```diff
@@ -33,6 +33,11 @@ public class LibraryDependency : IEquatable<LibraryDependency>
+        /// <summary>
+        /// Information regarding if the dependency is direct or transitive.  
+        /// </summary>
+        public LibraryDependencyReferenceType ReferenceType { get; set; } = LibraryDependencyReferenceType.Direct;
```
Feels like this adds duplication. 
TargetFrameworkInformation owns LibraryDependency and it owns CentralPackageVersions.
We explicitly wanted CentralPackageVersions to not be a part of LibraryDependency because, but now we are putting it back.

zivkan replied:
I agree that it feels like duplication. But when I was looking more closely (and debugging), I think it's needed with the current implementation, so it can detect the difference between direct dependencies and transitive dependencies (direct dependencies do not need to be added at the end of the graph walk, like the transitive dependencies do). But if I'm wrong and it can be removed, that'd be great.

nkolev92 replied:
Discussed on call.
My comments were stylistic and we have a consensus that it's ok to stay with the current implementation. There's not a great enough reason to change it.

nkolev92 commented on src/NuGet.Core/NuGet.ProjectModel/PackageSpecReferenceDependencyProvider.cs:
```diff
@@ -311,7 +311,7 @@ private List<LibraryDependency> GetDependenciesFromSpecRestoreMetadata(PackageSp
         }
 
-        private static List<LibraryDependency> GetSpecDependencies(
+        internal static List<LibraryDependency> GetSpecDependencies(
```
To honor @zivkan's comment about fake dependencies, there'd probably be a different call to get the central dependencies.

nkolev92 replied:
Same thing applies

nkolev92 reviewed (commented):
Similar concerns as @zivkan

nkolev92 commented on src/NuGet.Core/NuGet.ProjectModel/DependencyGraphSpec.cs:
```diff
@@ -308,7 +308,7 @@ public static DependencyGraphSpec Load(string path)
                         switch (propertyName)
                         {
-                           case "restore":
+                            case "restore":
```
nit: revert?

cristinamanum replied:
Reverted

nkolev92 commented on test/NuGet.Core.Tests/NuGet.Commands.Test/RestoreCommandTests.cs:
```diff
@@ -1416,6 +1417,215 @@ public async Task RestoreCommand_CentralVersion_ErrorWhenCentralPackageVersionFi
+        [Fact]
+        public async Task RestoreCommand_LogDowngradeWarningsOrErrorsAsync_ErrorWhenCpvmEnabled()
+        {
+            // Arange
```
```suggestion
            // Arrange
```

cristinamanum replied:
Updated.

nkolev92 commented on test/NuGet.Core.Tests/NuGet.Commands.Test/RestoreCommandTests.cs:
```diff
@@ -1416,6 +1417,215 @@ public async Task RestoreCommand_CentralVersion_ErrorWhenCentralPackageVersionFi
+        {
+            // Arange
+            // create graph with a downgrade
+            string centralPackageName = "D";
```
I don't want to be too picky on tests but fwiw, this is not in line with our guideline
See 10 from https://github.com/NuGet/NuGet.Client/blob/dev/docs/coding-guidelines.md

cristinamanum replied:
Updated

nkolev92 commented on test/NuGet.Core.Tests/NuGet.Commands.Test/RestoreCommandTests.cs:
```diff
@@ -1416,6 +1417,215 @@ public async Task RestoreCommand_CentralVersion_ErrorWhenCentralPackageVersionFi
+        }
+
+        [Fact]
+        public async Task RestoreCommand_DowngradeIsErrorWhen_DowngradedByCentralTransitiveDependency()
```
Discussed during the meeting, we can add a bunch more tests scenarios.
We should probably create an issue for that.

cristinamanum replied:
Added https://github.com/NuGet/Home/issues/9318

nkolev92 reviewed (approved):
These 4 comments + the 2 leftover from the previews review. 

Looks good beyond that. 
Sorry for the delay in the review. 
Happy to chat through the feedback before merging.

cristinamanum committed:
Feedback

cristinamanum committed:
Test updates'
